doc/hacking/hacking.adoc

```diff
@@ -1045,6 +1045,145 @@ The code is generated as a list, in the `*code-instr*` global variable
 in the `STKLOS-COMPILER` module. The procedure `emit` conses one more
 instruction on the code (which will later be reversed, of course)
 
+==== Standard and new special forms
+
+The set of special forms in STklos is extendable, and the whole
+mechanism is illustrated in this section with an example.
+
+The special form `if` is a macro. It is defined in `lib/runtime-macros.stk` as
+
+[source, scheme]
+----
+(define-macro (if . args)            `(#%if ,@args))
+----
+
+The `#%if` to which it exapands into is an *internal compiler call*
+(named `icall` in STklos). It looks like a Scheme symbol, but symbols
+cannot start with the sharp sign (`#`), so one cannot use this as a symbol name:
+
+[source, scheme]
+----
+stklos> (define #%if 10)
+**** Error:
+compiler-error: define: bad variable name #%if
+----
+
+However, `icall` objects are self-evaluating, just like symbols:
+
+[source, scheme]
+----
+stklos> #%if
+#%if
+----
+
+This makes it possible to redefine the symbols that usually are names of
+special forms, like `if`, `cond`, `quote` etc.
+
+When the compiler looks at the form `(#%if bool something)`,
+it will recognize that the `car` of the form is an `icall` object,
+and automatically call a procedure that compiles it.
+
+Each `icall` object has
+* a name (a Scheme symbol)
+* a "function" (actually the Scheme procedure that compiles this form, just
+  like the others in `compiler.stk`). The function field may be set to `#f`.
+
+If the function field of an `icall` object is `#f`, then the function called to
+compile it is `compile-%%NAME`, where `NAME` is the name of the `icall`
+object.
+
+[source, scheme]
+----
+(%icall? #%if)         => #t
+(%icall-name #%if)     => if
+(%icall-function #%if) => #f
+----
+
+Since the function attached to `#%if` is `#f`, then `compile-%%if` (which is
+defined inside the `STKLOS-COMPILER` module) is called.
+
+[source, scheme]
+----
+(if bool something)
+  ~~> (#%if bool something)
+    ~~> (compile-%%if '((#%if bool something) #f #f)     ;; if env and tail are both #f
+----
+
+We now implement a special form `twice`, that compiles the code inside it twice.
```

Comment:
I think that there is misunderstanding here. Internal icalls are useful for structure which need to produce code (e.g. new instructions), or need to access to the internals of the compiler. `Twice` could be easily implemented by a macro:

```scheme
( define-macro (twice . args) `(begin ,@args ,@args))
```
I think, that the example with `twice` is confusing and should not be here.